coreneuron/sim/scopmath/crout_thread.hpp

```diff
@@ -0,0 +1,190 @@
+                sum += a[irow][scopmath_crout_ix(k)] * a[krow][scopmath_crout_ix(r)];
+            }
+            a[irow][scopmath_crout_ix(r)] -= sum;
+        }
+
+        /* Find row containing the pivot in the rth column */
+        int pivot = perm[scopmath_crout_ix(r)];
+        double equil_1 = fabs(a[pivot][scopmath_crout_ix(r)] / rowmax[scopmath_crout_ix(pivot)]);
+        for (int i = r + 1; i < n; i++) {
+            int irow = perm[scopmath_crout_ix(i)];
+            double equil_2 = fabs(a[irow][scopmath_crout_ix(r)] / rowmax[scopmath_crout_ix(irow)]);
+            if (equil_2 > equil_1) {
+                /* make irow the new pivot row */
+
+                pivot = irow;
+                save_i = i;
+                equil_1 = equil_2;
+            }
+        }
+
+        /* Interchange entries in permutation vector if necessary */
+        if (pivot != perm[scopmath_crout_ix(r)]) {
+            perm[scopmath_crout_ix(save_i)] = perm[scopmath_crout_ix(r)];
+            perm[scopmath_crout_ix(r)] = pivot;
+        }
+
+        /* Check that pivot element is not too small */
+        if (fabs(a[pivot][scopmath_crout_ix(r)]) < ROUNDOFF)
+            return SINGULAR;
+
+        /*
+         * Operate on row in rth position.  This produces the upper
+         * triangular matrix whose diagonal elements are assumed to be unity.
+         * This matrix is used in the back substitution algorithm.
+         */
+        for (int j = r + 1; j < n; j++) {
+            double sum = 0.0;
+            for (int k = 0; k < r; k++) {
+                int krow = perm[scopmath_crout_ix(k)];
+                sum += a[pivot][scopmath_crout_ix(k)] * a[krow][scopmath_crout_ix(j)];
+            }
+            a[pivot][scopmath_crout_ix(j)] = (a[pivot][scopmath_crout_ix(j)] - sum) /
+                                             a[pivot][scopmath_crout_ix(r)];
+        }
+    }
+    return SUCCESS;
+}
+
+/**
+ * Performs forward substitution algorithm to transform the constant vector in
+ * the linear simultaneous equations to be consistent with the factored matrix.
+ * Then performs back substitution to find the solution to the simultaneous
+ * linear equations.
+ *
+ * @param n number of rows of the matrix
+ * @param a double precision matrix containing the factored matrix of
+ *          coefficients of the linear equations
+ * @param b vector of function values
+ * @param perm permutation vector to store row interchanges
+ * @param[out] p[y[i]] contains the solution vector
+ */
+inline void nrn_scopmath_solve_thread(int n,
+                                      double** a,
+                                      double* b,
+                                      int* perm,
+                                      double* p,
+                                      int* y,
+                                      _threadargsproto_) {
+    /* Perform forward substitution with pivoting */
+    // if (y) { // pgacc bug. nullptr on cpu but not on GPU
```

Comment:
Just a note: looking at bed333d4a6e1088f3e224efdfe04d498362db6a7, looks like this was relevant in ~2016. May be this is not required anymore.

Reply:
Maybe not, but apparently it hasn't been causing problems so I didn't think I needed to touch it...

Reply:
yeah, we don't need to touch it in this PR.